### PR TITLE
V1.4 Fees

### PR DESCRIPTION
**Pull request by lucdx**
- Fixed all bugs raised during PE1

- new PaidFeesCommand to make it easier for Admins to set users who have paid their fees instead of using EditFeesCommand which requires a date.

### CODE REVIEW

**Comment by meisbokai** [comment]
the snake case requires the first word to be small letters

**Comment by micklwl** [comment]
The test names should be
featureUnderTest_testScenario_expectedBehavior()

File: src/seedu/addressbook/commands/fees/EditFeesCommand.java [hunk 1]
```diff
@@ -33,6 +33,12 @@
...
     public EditFeesCommand(int index, String fees, String date) throws IllegalValueException {
         setTargetIndex(index, ObjectTargeted.PERSON);
+        if (!isValidDate(date) && !"0".equals(date)) {
+            throw new IllegalValueException(MESSAGE_DATE_CONSTRAINTS);
+        }
+        if ("0.00".equals(fees)) {
+            throw new IllegalValueException(MESSAGE_FEES_VALUE_CONSTRAINTS);
+        }
         this.fees = new Fees(fees, date);
     }
```
Comment (0WN463) [inline src/seedu/addressbook/commands/fees/EditFeesCommand.java]: This looks quite potato. Cause you'll allow things like "Garbage" as your fees.
fees should be a double which you convert from String to double.
Then you can also check if it is converted properly and that it is in the $ format, so not things like $0.3333.
If it isn't formatted properly, throw a error message
But it'll require quite a lot of changes to make it clean though

File: src/seedu/addressbook/commands/fees/PaidFeesCommand.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    public static final String MESSAGE_SUCCESS = "Fees paid: %1$s";
+
+    /**
+     * Constructor used for Privileges
+     * Command constructed has no functionality
+     * */
+    //public PaidFeesCommand() {
+    // Does nothing
+    //}
+
```
Comment (0WN463) [inline src/seedu/addressbook/commands/fees/PaidFeesCommand.java]: If not needed, then remove

File: src/seedu/addressbook/common/Utils.java [hunk 3]
```diff
@@ -64,4 +67,24 @@ public static boolean isValidDate(String value) {
...
+            main1.append(o1.getFees().duedate.substring(6, 10));
+            main1.append(o1.getFees().duedate.substring(3, 5));
+            main1.append(o1.getFees().duedate.substring(0, 2));
+            main2.append(o2.getFees().duedate.substring(6, 10));
+            main2.append(o2.getFees().duedate.substring(3, 5));
+            main2.append(o2.getFees().duedate.substring(0, 2));
+            return main1.toString().compareTo
+                    (main2.toString());
+        }
+    }
```
Comment (0WN463) [inline src/seedu/addressbook/common/Utils.java]: Something is wrong when even I don't understand this code.
Also, duedate should not be public

File: src/seedu/addressbook/data/person/Fees.java [hunk 4]
```diff
@@ -9,13 +9,21 @@
...
     public final String value;
     public final String duedate;
+    private boolean isEdited = false;
+
+    //TODO hide the fees when it's first initialised as 00-00-0000
+    // new command for when they paid fees, fully, thus reinitiliasing as 00-00-0000
+    // so that it doesnt fail valid date check in edit fees
+    // therefore fees will always either be 00-00-0000 or a valid date,
+    // and will always be hidden when its 00-00-0000
+    // this way is easier as dont need to change all the adapted person and stuff
```
Comment (0WN463) [inline src/seedu/addressbook/data/person/Fees.java]: What's the point of TODO if this is the final iteration?

File: src/seedu/addressbook/data/person/ReadOnlyPerson.java [hunk 5]
```diff
@@ -154,9 +156,16 @@ default String getAsTextShowFee() {
...
+                getName());
         builder.append(stringChain);
+        if (getFees().isEdited()) {
+            builder.append(Formatter.getPrintableString(true, getFees()));
+        } else {
+            //builder.append("{private Fees: ");
+            //builder.append(getFees().value);
+            //builder.append(" / }\n");
+            builder.append("No Fees owed!\n");
+        }
```
Comment (0WN463) [inline src/seedu/addressbook/data/person/ReadOnlyPerson.java]: Same, if not needed, remove

File: test/java/seedu/addressbook/TestDataHelper.java [hunk 6]
```diff
@@ -53,9 +53,9 @@ public Exam math() throws Exception {
...
 
     /** Test fees for testing**/
-    public Fees fees () throws Exception {
-        String test = "123.45";
-        String testdate = "01-01-2018";
+    public Fees fees (int seed) throws Exception {
+        String test = seed + ".45";
+        String testdate = "01-0" + seed + "-2018";
         return new Fees(test, testdate);
     }
```
Comment (0WN463) [inline test/java/seedu/addressbook/TestDataHelper.java]: Bad variable naming. What is test?

File: src/seedu/addressbook/commands/fees/EditFeesCommand.java (hunk 1 above)
Comment (lucdx) [reply src/seedu/addressbook/commands/fees/EditFeesCommand.java]: Actually the regex alr checks for correct values as Strings, if u look under the Fees class. It's just that idk how to make 1 exception for the 0 value, that's why this is here.

So things like "Garbage" or "0.33333333" won't go through.....

File: src/seedu/addressbook/commands/fees/PaidFeesCommand.java (hunk 2 above)
Comment (lucdx) [reply src/seedu/addressbook/commands/fees/PaidFeesCommand.java]: Oki!

File: src/seedu/addressbook/data/person/Fees.java (hunk 4 above)
Comment (lucdx) [reply src/seedu/addressbook/data/person/Fees.java]: Oh it's done alr sorry let me remove it

File: src/seedu/addressbook/common/Utils.java (hunk 3 above)
Comment (lucdx) [reply src/seedu/addressbook/common/Utils.java]: u mean the comparator or this particular return thing?

File: src/seedu/addressbook/data/person/ReadOnlyPerson.java (hunk 5 above)
Comment (lucdx) [reply src/seedu/addressbook/data/person/ReadOnlyPerson.java]: Yup!

File: test/java/seedu/addressbook/TestDataHelper.java (hunk 6 above)
Comment (lucdx) [reply test/java/seedu/addressbook/TestDataHelper.java]: Oki shall rename it!

File: src/seedu/addressbook/commands/fees/EditFeesCommand.java (hunk 1 above)
Comment (0WN463) [reply src/seedu/addressbook/commands/fees/EditFeesCommand.java]: It still weird that your fees value is stored as a string instead of a number data type

Comment (lucdx) [reply src/seedu/addressbook/commands/fees/EditFeesCommand.java]: Had this conversation w Mick early on during the Sem and I forgot what we talked about but we agreed on using Strings

Comment (0WN463) [reply src/seedu/addressbook/commands/fees/EditFeesCommand.java]: Fine, I yield

File: src/seedu/addressbook/common/Utils.java (hunk 3 above)
Comment (0WN463) [reply src/seedu/addressbook/common/Utils.java]: what does all the append/substring mean?

Comment (lucdx) [reply src/seedu/addressbook/common/Utils.java]: its removing all the dashes in the DD-MM-YYYY and appending into a string for YYYYMMDD to sort

Comment (0WN463) [reply src/seedu/addressbook/common/Utils.java]: So it'll be good to pull it into its own function called convertToString or something

Comment (lucdx) [reply src/seedu/addressbook/common/Utils.java]: Ok I'll convert it! Should i put it under Utils also?

Comment (0WN463) [reply src/seedu/addressbook/common/Utils.java]: Then you can call it in your UniquePersonList also

Comment (lucdx) [reply src/seedu/addressbook/common/Utils.java]: Gotcha

**Pull request merged**